### PR TITLE
add composite img and ktx image loader

### PR DESCRIPTION
**Pull request by ZihaoTu**

### CODE REVIEW

File: src/CMakeLists.txt [hunk 1]
```diff
@@ -14,8 +14,7 @@ find_package(PNG REQUIRED)
 
 set(IMGPP_HEADER
   include/imgpp/algorithms.hpp
-  include/imgpp/bcdesc.hpp
-  include/imgpp/bchelper.hpp
+  include/imgpp/textureformat.hpp
```
Comment (chubei-oppen) [inline src/CMakeLists.txt]: bump version

File: src/include/imgpp/compositeimg.hpp [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+  void AddBCROI(std::vector<BCImgROI> bcimg_rois) {
+    bcimg_rois_ = std::move(bcimg_rois);
+  }
+
+  const BCImgROI &BCROI(uint32_t level, uint32_t layer, uint32_t face) const {
+    return bcimg_rois_[level * layers_ * faces_ + layer * faces_ + face];
+  }
+
+  void SetInfo(std::string &&key, std::string &&value) {
+    customized_info_.insert_or_assign(key, value);
```
Comment (chubei-oppen) [inline src/include/imgpp/compositeimg.hpp]: std::move(key), std::move(value)

File: src/include/imgpp/compositeimg.hpp [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+  }
+
+  void SetInfo(const std::string &key, const std::string &value) {
```
Comment (chubei-oppen) [inline src/include/imgpp/compositeimg.hpp]: just pass by value?

File: src/src/ktximage.cpp [hunk 4]
```diff
@@ -0,0 +1,325 @@
...
+    header.pixel_height = IsTarget1d(target) ? 0 : roi.Height();
+    header.pixel_depth = IsTarget3d(target) ? roi.Depth() : 0;
+  }
+  header.number_of_array_elements = IsTargetArray(target) ? composite_img.Layers() : 0;
+  header.number_of_faces = IsTargetCube(target) ? 6 : 1;
+  header.number_of_mipmap_levels = composite_img.NeedGenerateMipmaps() ? 0 : composite_img.Levels();
+  offset += sizeof(KTXHeader);
+
+  const auto &kv_map = composite_img.Info();
+  if (!kv_map.empty()) {
```
Comment (chubei-oppen) [inline src/src/ktximage.cpp]: unnecessary if

File: src/src/ktximage.cpp [hunk 5]
```diff
@@ -0,0 +1,325 @@
...
+}
+
+bool WriteKTX(const char *fn, const CompositeImg &composite_img, bool bottom_first) {
+  if (bottom_first) {
+    return false;
+  }
+  if (composite_img.Format() == FORMAT_UNDEFINED) {
+    return false;
+  }
+  uint32_t total_size = ComputeKTXStorageSize(composite_img);
```
Comment (chubei-oppen) [inline src/src/ktximage.cpp]: why allocate the whole file?

File: src/src/ktximage.cpp [hunk 6]
```diff
@@ -0,0 +1,325 @@
...
+      }
+      for (uint32_t layer = 0; layer < composite_img.Layers(); ++layer) {
+        for (uint32_t face = 0; face < composite_img.Faces(); ++face) {
+          std::memcpy(data.data() + offset, composite_img.ROI(level, layer, face).GetData(), face_size);
+          offset += face_size;
+        }
+      }
+    }
+  }
+  std::ofstream out(fn, std::ios::binary);
```
Comment (chubei-oppen) [inline src/src/ktximage.cpp]: try open file before doing all the heavy work

File: src/src/ktximage.cpp [hunk 7]
```diff
@@ -0,0 +1,325 @@
...
+  if (!out.good()) {
+    return false;
+  }
+  out.write((char*)data.data(), data.size());
+  return true;
```
Comment (chubei-oppen) [inline src/src/ktximage.cpp]: return out.good()

File: src/src/ktxloadertest.cpp [hunk 8]
```diff
@@ -0,0 +1,19 @@
+#include <iostream>
+#include <vector>
+#include <string>
+#include <imgpp/imgpp.hpp>
+#include <imgpp/loaders.hpp>
+#include <imgpp/loadersext.hpp>
+#include <imgpp/compositeimg.hpp>
+
+int main(int argc, char* argv[]) {
```
Comment (chubei-oppen) [inline src/src/ktxloadertest.cpp:9]: ctest test can't be implemented like this

File: src/include/imgpp/gl.hpp [hunk 9]
```diff
@@ -0,0 +1,334 @@
...
+  KTX
+};
+
+struct FormatDesc {
+  InternalFormat internal;
+  ExternalFormat external;
+  TypeFormat type;
+};
+
+inline InternalFormat TranslateFromES20(InternalFormat internal) {
```
Comment (shenfy) [inline src/include/imgpp/gl.hpp]: move functions to cpp file in an anonymous namespace

File: src/include/imgpp/compositeimg.hpp [hunk 10]
```diff
@@ -0,0 +1,119 @@
+#ifndef IMGPP_COMPOSITEIMG_HPP
+#define IMGPP_COMPOSITEIMG_HPP
+
+#include <string>
```
Comment (shenfy) [inline src/include/imgpp/compositeimg.hpp]: remove

File: src/include/imgpp/compositeimg.hpp [hunk 11]
```diff
@@ -0,0 +1,119 @@
...
+public:
+  CompositeImg() {
+  };
+  CompositeImg(TextureDesc desc, uint32_t levels, uint32_t layers, uint32_t faces)
+    : texture_desc_(desc), levels_(levels), layers_(layers), faces_(faces) {
+  }
+  ~CompositeImg() {
+  };
+
+  void SetTextureDesc(TextureDesc desc) {
```
Comment (shenfy) [inline src/include/imgpp/compositeimg.hpp]: use reference or move to avoid copying twice.

File: src/include/imgpp/compositeimg.hpp [hunk 12]
```diff
@@ -0,0 +1,119 @@
...
+    texture_desc_ = desc;
+  }
+
+  const TextureDesc &GetTextureDesc() const {
```
Comment (shenfy) [inline src/include/imgpp/compositeimg.hpp]: change name to TextureDesc()

File: src/include/imgpp/compositeimg.hpp [hunk 13]
```diff
@@ -0,0 +1,119 @@
...
+    if (texture_desc_.format != FORMAT_UNDEFINED) {
+      levels_ = levels;
+      layers_ = layers;
+      faces_ = faces;
+      align_bytes_ = align_bytes;
+      img_rois_.resize(levels * layers * faces);
+      uint32_t pitch = ImgROI::CalcPitch(width, c, bpc, align_bytes);
+      ImgROI roi(nullptr, width, height, depth, c, bpc,
+        pitch, pitch * height, is_float, is_signed);
+      img_rois_[0] = roi;
```
Comment (shenfy) [inline src/include/imgpp/compositeimg.hpp]: move or emplace_back

File: src/include/imgpp/compositeimg.hpp [hunk 14]
```diff
@@ -0,0 +1,119 @@
...
+    }
+  }
+
+  void SetSize(uint32_t levels, uint32_t layers, uint32_t faces,
```
Comment (shenfy) [inline src/include/imgpp/compositeimg.hpp]: change method name, be explicit about this being a BC image

File: src/include/imgpp/compositeimg.hpp [hunk 15]
```diff
@@ -0,0 +1,119 @@
...
+#include <imgpp/imgpp_bc.hpp>
+#include <imgpp/texturedesc.hpp>
+
+namespace imgpp {
+
+class CompositeImg {
+public:
+  CompositeImg() {
+  };
+  CompositeImg(TextureDesc desc, uint32_t levels, uint32_t layers, uint32_t faces)
```
Comment (shenfy) [inline src/include/imgpp/compositeimg.hpp]: invalid constructor! ROIs not allocated!

File: src/include/imgpp/compositeimg.hpp [hunk 16]
```diff
@@ -0,0 +1,119 @@
...
+    }
+  }
+
+  void SetData(uint8_t *data, uint32_t level, uint32_t layer, uint32_t face) {
+    if (IsCompressed()) {
+      uint32_t width = std::max(bcimg_rois_[0].Width() >> level, 1u);
+      uint32_t height = std::max(bcimg_rois_[0].Height() >> level, 1u);
+      uint32_t depth = std::max(bcimg_rois_[0].Depth() >> level, 1u);
+      BCImgROI bcimg_roi(data, texture_desc_.format, width, height, depth);
+      bcimg_rois_[level * layers_ * faces_ + layer * faces_ + face] = bcimg_roi;
```
Comment (shenfy) [inline src/include/imgpp/compositeimg.hpp]: use move or emplace_back

File: src/include/imgpp/compositeimg.hpp [hunk 17]
```diff
@@ -0,0 +1,119 @@
...
+    } else {
+      uint32_t width = std::max(img_rois_[0].Width() >> level, 1u);
+      uint32_t height = std::max(img_rois_[0].Height() >> level, 1u);
+      uint32_t depth = std::max(img_rois_[0].Depth() >> level, 1u);
+      uint32_t pitch = ImgROI::CalcPitch(width, img_rois_[0].Channel(), img_rois_[0].BPC(), align_bytes_);
+      ImgROI roi = ImgROI(data, width, height, depth,
+          img_rois_[0].Channel(), img_rois_[0].BPC(), pitch, pitch * height,
+          img_rois_[0].IsFloat(), img_rois_[0].IsSigned());
+      img_rois_[level * layers_ * faces_ + layer * faces_ + face] = roi;
```
Comment (shenfy) [inline src/include/imgpp/compositeimg.hpp]: use move or emplace_back

File: src/include/imgpp/compositeimg.hpp [hunk 18]
```diff
@@ -0,0 +1,119 @@
...
+    }
+  }
+
+  void AddBuffer(imgpp::ImgBuffer buffer) {
+    buffers_.push_back(std::move(buffer));
+  }
+
+  const ImgROI &ROI(uint32_t level, uint32_t layer, uint32_t face) const {
```
Comment (shenfy) [inline src/include/imgpp/compositeimg.hpp:76]: why not provide access to non-const ROI?

File: src/include/imgpp/gl.hpp [hunk 19]
```diff
@@ -0,0 +1,302 @@
+#ifndef IMGPP_GL_HPP
+#define IMGPP_GL_HPP
+
+#include <imgpp/texturedesc.hpp>
+
+namespace imgpp { namespace gl {
+enum InternalFormat: uint16_t {
```
Comment (shenfy) [inline src/include/imgpp/gl.hpp]: move all GL macro definitions into cpp anonymous namespace.

File: src/include/imgpp/gl.hpp [hunk 20]
```diff
@@ -0,0 +1,302 @@
...
+};
+
+struct FormatDesc {
+  InternalFormat internal;
+  ExternalFormat external;
+  BaseInternalFormat base_internal;
+  TypeFormat type;
+};
+
+TextureFormat Find(InternalFormat internal, ExternalFormat external,
```
Comment (shenfy) [inline src/include/imgpp/gl.hpp]: change name to TranslateFromGL

File: src/include/imgpp/gl.hpp [hunk 21]
```diff
@@ -0,0 +1,302 @@
...
+  BaseInternalFormat base_internal, TypeFormat type);
+FormatDesc Translate(TextureFormat format, Profile profile);
```
Comment (shenfy) [inline src/include/imgpp/gl.hpp]: change name to Translate2GL

File: src/include/imgpp/gl.hpp [hunk 22]
```diff
@@ -0,0 +1,302 @@
...
+
+enum Profile: uint8_t {
+  ES20, // no sRGB
+  ES30, // has swizzle
+  GL32,
+  GL33, // has swizzle
+  KTX
+};
+
+struct FormatDesc {
```
Comment (shenfy) [inline src/include/imgpp/gl.hpp]: change type and names: e.g. uint16_t internal_format

**Review by shenfy — CHANGES_REQUESTED** [review]

File: src/src/ktximage.cpp [hunk 23]
```diff
@@ -0,0 +1,317 @@
...
+  desc.target = texture_target;
+  desc.mipmap = ktx_header.number_of_mipmap_levels != 1;
+  composite_img.SetTextureDesc(desc);
+  std::array<uint32_t, 3> original_extent = {
+    std::max(ktx_header.pixel_width, 1u),
+    std::max(ktx_header.pixel_height, 1u),
+    std::max(ktx_header.pixel_depth, 1u)
+  };
+  // Parse user-defined key-value data
+  int64_t kv_left = ktx_header.bytes_of_key_value_data;
```
Comment (chubei-oppen) [inline src/src/ktximage.cpp:140]: custom_data.clear();

**Pull request merged**